File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
import heapq
import sys

def dijkstra(start, distance, graph):
    q = []
    heapq.heappush(q,(0,start))
    distance[start] = 0
    while q:
        dist, now = heapq.heappop(q)
        # print("node : ",now)
        if distance[now] < dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q,(cost,i[0]))

                
def solution(n, vertex):
    answer = 0
    vertex.sort()
    INF = int(1e9)
    graph = [[] for i in range(n+1)]
    distance = [INF]*(n+1)
    answer = 0
    
    
    for i in range(len(vertex)):
        graph[vertex[i][0]].append((vertex[i][1],1))
        graph[vertex[i][1]].append((vertex[i][0],1))
    
    dijkstra(1, distance, graph)
    max = distance[1]
    for i in range(len(distance)):
        if distance[i] != INF and max <= distance[i]:
            max = distance[i]
    for i in range(len(distance)):
        if distance[i] == max:
            answer += 1
    
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        int[] dist = new int[n+1];
        boolean[] visited = new boolean[n+1];
        Arrays.fill(dist, 0);
        
        for(int i = 0; i <= n; i++){
            graph.add(new ArrayList<Integer>());
        }
        
        
        for(int i = 0; i < edge.length; i++){
            graph.get(edge[i][0]).add(edge[i][1]);
            graph.get(edge[i][1]).add(edge[i][0]);
        }
        
        for(ArrayList<Integer> a : graph){
            Collections.sort(a);
        }
        
        LinkedList<Integer> q = new LinkedList<>();
        
        visited[1] = true;
        for(int a : graph.get(1)){
            q.add(a);
            visited[a] = true;
            dist[a] = 1;
        }
        
        while(q.size() > 0) {
            int node = q.pop();
            int distance = dist[node];
            for(int a : graph.get(node)){
                if (!visited[a]) {
                    q.add(a);
                    dist[a] = distance + 1;
                    visited[a] = true;
                }
            }
        }
        
        int max = 0;
        
        for(int i = 0; i < dist.length; i++){
            if (dist[i] > max) {
                max = dist[i];
            }
        }
        
        for(int i = 0; i < dist.length; i++){
            if (dist[i] == max) {
                answer += 1;
            }
        }
        
        
        return answer;
    }
}